scripts/lib/dependencies.ts

```diff
@@ -128,45 +100,207 @@ export async function writeDependencyVersions(path: string, depVersions: Depende
   await fs.promises.writeFile(path, rawContents, { encoding: 'utf-8' });
 }
 
+/**
+ * A dependency is some binary that we need to track.  Generally this is some
+ * third-party software, but it may also be things we build in an external
+ * repository, or some binary we build from them.
+ */
 export interface Dependency {
-  name: string,
+  /** The name of this dependency. */
+  readonly name: string,
   /**
    * Other dependencies this one requires.
    * This must be in the form <name>:<platform>, e.g. "kuberlr:linux"
    */
   dependencies?: (context: DownloadContext) => string[],
+  /**
+   * Download this dependency.  Note that for some dependencies, this actually
+   * builds from source.
+   */
   download(context: DownloadContext): Promise<void>
-  // Returns the available versions of the Dependency.
-  // Includes prerelease versions if includePrerelease is true.
-  getAvailableVersions(includePrerelease?: boolean): Promise<string[] | AlpineLimaISOVersion[]>
-  // Returns -1 if version1 is higher, 0 if version1 and version2 are equal,
-  // and 1 if version2 is higher.
-  rcompareVersions(version1: string | AlpineLimaISOVersion, version2: string | AlpineLimaISOVersion): -1 | 0 | 1
 }
 
 /**
- * A Dependency that is hosted in a GitHub repo.
+ * A VersionedDependency is a {@link Dependency} where we track a version and
+ * can be automatically upgraded (i.e. a pull request made to bump the version).
  */
-export interface GitHubDependency extends Dependency {
-  githubOwner: string
-  githubRepo: string
-  // Converts a version (of the format that is stored in dependencies.yaml)
-  // to a tag that is used in a GitHub release.
-  versionToTagName(version: string | AlpineLimaISOVersion): string
-}
+export abstract class VersionedDependency implements Dependency {
+  abstract readonly name: string;
+  abstract download(context: DownloadContext): Promise<void>;
+  /**
+   * Returns the available versions of the Dependency.
+   */
+  abstract getAvailableVersions(): Promise<Version[]>;
 
-export function IsGitHubDependency(dependency: Dependency): dependency is GitHubDependency {
-  if (!('githubOwner' in dependency) || typeof dependency.githubOwner !== 'string') {
-    return false;
+  /** The current version. */
+  abstract currentVersion: Promise<Version>;
+
+  /** The newest version that can be upgraded to. */
+  get latestVersion(): Promise<Version> {
+    return (async() => {
+      const availableVersions = await this.getAvailableVersions();
+
+      return availableVersions.reduce((version1, version2) => {
+        return this.rcompareVersions(version1, version2) < 0 ? version1 : version2;
```

Comment:
Wait, doesn't this select the oldest version instead of the the latest due to inverted logic?
Shouldn't it be: 
```
return availableVersions.reduce((v1, v2) => {
  return this.rcompareVersions(v1, v2) > 0 ? v1 : v2;
});
```